readthedocs/vcs_support/backends/git.py

```diff
@@ -68,6 +87,153 @@ def update(self):
+            # was an assumption in the .branches property that iterates over everything in
+            # refs/remotes/origin. We might still keep this pattern as it seems to be the most solid
+            # convention for storing a remote:local refspec mapping.
+            return (
+                f"refs/heads/{self.version_identifier}:refs/remotes/origin/"
+                f"{self.version_identifier}"
+            )
+        # Tags
+        if self.version_type == TAG and self.verbose_name:
+            # A "stable" tag is automatically created with Version.machine=True,
+            # denoting that it's not a tag that really exists.
+            # Because we don't know if it originates from the default branch or some
+            # other tagged release, we will fetch everything.
+            log.info(
+                "Is this an auto-created version?", version_machine=self.version_machine
+            )
+            if self.version_machine:
+                return None
+            return f"refs/tags/{self.verbose_name}:refs/tags/{self.verbose_name}"
+
+        if self.version_type == EXTERNAL:
+            # TODO: We should be able to resolve this without looking up in oauth registry
+            git_provider_name = self.project.git_provider_name
+
+            # Remote reference for Git providers where pull request builds are supported
+            if git_provider_name == GITHUB_BRAND:
+                return GITHUB_PR_PULL_PATTERN.format(id=self.verbose_name)
+            if self.project.git_provider_name == GITLAB_BRAND:
+                return GITLAB_MR_PULL_PATTERN.format(id=self.verbose_name)
+
+    def clone_ng(self):
+        # TODO: This seems to be legacy that can be removed.
+        #  If the repository is already cloned, we don't do anything.
+        #  It seems to originate from when a cloned repository was cached on disk,
+        #  and so we can call call .update() several times in the same build.
+        if self.repo_exists():
+            return
+
+        # TODO: This seems to be legacy that can be removed.
+        #  There shouldn't be cases where we are asked to
+        #  clone the repo in a non-clean working directory.
+        #  The prior call to repo_exists() will return if a repo already exist with
+        #  unclear guarantees about whether that even needs to be a fully consistent clone.
+        self.make_clean_working_dir()
+
+        # --no-checkout: Makes it explicit what we are doing here. Nothing is checked out
+        #                until it's explicitly done.
+        # --depth 1: Shallow clone, fetch as little data as possible.
+        cmd = ["git", "clone", "--depth", "1"]
+
+        # Everything EXCEPT unnamed branch builds can have "--no-checkout"
+        # This is the case for building a manually imported project for the first time.
+        if not (self.version_type == BRANCH and not self.version_identifier):
+            cmd.append("--no-checkout")
```

Comment:
I'm not sure to understand this comment here, it would be good if we can expand the comment in a clearer way.

Also, it seems this case is related to the "the first build of the project" (where we don't have an identifier for the version). Besides, in this case we do a `git clone` (and a checkout in the same command); I think we should not do the `git fetch` and `git checkout` later. So, we probably need to save this somehow (maybe a an instance variable) and avoid calling those other commands in this case.

This has to be pretty clearly explained because it's super hard to realize and follow all these cases. So, good comments are appreciated here.

Reply:
I've removed the `git fetch` operation as well now!

![image](https://github.com/readthedocs/readthedocs.org/assets/374612/c3a63f25-434f-4528-8b6d-ff807dd2155d)